File: libs/js-shared/base-workflow-frontend/src/lib/domain/definition/artifact-definition.ts
import { BaseEntity } from '@processpuzzle/base-entity';

/**
 * Frontend model of `ArtifactDefinition` — something a workflow produces or consumes whose
 * *lifecycle* is worth tracking, authored once per tenant under `/artifacts`.
 *
 * The tenant's catalog of *everything a task may touch*, and the reason a task's `inputs` and
 * `outputs` are plain ids: {@link artifactType} already says whether the thing is an entity, a
 * document or a widget, so there is nothing left for a per-reference type to add. An artifact may also
 * name a base-state machine, which is what makes its lifecycle trackable.
 */

/**
 * Mirrors the contract's `ArtifactType`: SPEM's kinds of work product, plus `ENTITY` for one whose data
 * is a base-entity instance. Names the kind, not the schema — every value here describes an artifact
 * definition, and this field says which kind of one it is. {@link ArtifactDefinition.artifactTypeId}
 * then names the concrete document, entity or widget.
 *
 * These are the contract's three values. The enum carried SPEM's older `ARTIFACT` / `DELIVERABLE` /
 * `OUTCOME` beside `ENTITY` until this revision, which left the form offering three values the backend
 * rejects and no way at all to author the `DOCUMENT` and `WIDGET` the seed data uses.
 */
export enum ArtifactType {
  DOCUMENT = 'DOCUMENT',
  ENTITY = 'ENTITY',
  WIDGET = 'WIDGET',
}

export class ArtifactDefinition implements BaseEntity {
  id: string;
  name: string;
  description?: string;
  artifactType: ArtifactType | undefined;
  /**
   * Id of the concrete thing this artifact *is*: a base-entity entity name when
   * {@link artifactType} is `ENTITY`, a base-document document name when `DOCUMENT`, a widget name
   * when `WIDGET`.
   */
  artifactTypeId?: string;
  /** Id of the base-state machine governing this artifact's lifecycle. */
  stateMachineId?: string;
  // region server-assigned
  version: number | undefined;
  createdAt: string | undefined;
  updatedAt: string | undefined;
  // endregion

  constructor(init: Partial<ArtifactDefinition> = {}) {
    this.id = init.id ?? '';
    this.name = init.name ?? '';
    this.description = init.description;
    this.artifactType = init.artifactType;
    this.artifactTypeId = init.artifactTypeId;
    this.stateMachineId = init.stateMachineId;
    this.version = init.version;
    this.createdAt = init.createdAt;
    this.updatedAt = init.updatedAt;
  }
}

File: libs/java-shared/base-workflow-backend/src/main/java/com/processpuzzle/workflow/definition/domain/ArtifactDefinition.java
package com.processpuzzle.workflow.definition.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.Id;
import jakarta.persistence.IdClass;
import jakarta.persistence.Table;
import jakarta.persistence.Version;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import lombok.ToString;

/**
 * SPEM's work product, under the name the platform uses for it: everything a task reads or writes
 * is one of these, because an artifact is also where the lifecycle lives — {@link #artifactType}
 * says what kind of thing it is and {@link #stateMachineId} which base-state machine governs it.
 *
 * <p>A catalog entity like {@link RoleDefinition}: the same artifact is produced or consumed by
 * tasks of several workflows, so it is defined once per organization and referenced from every
 * {@link Workflow#getArtifacts()} that involves it, through an {@link ArtifactUse}.
 */
@Getter
@Setter
@Builder
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = false, of = {"orgKey", "id"})
@ToString
@Entity
@Table(name = "workflow_artifact_definition")
@IdClass(ArtifactDefinitionKey.class)
public class ArtifactDefinition extends com.processpuzzle.workflow.common.Auditable {

    @Id
    @Column(name = "org_key", nullable = false)
    private String orgKey;

    /** Artifact code, chosen by the author and unique per organization. */
    @Id
    @Column(nullable = false)
    private String id;

    @Column(nullable = false)
    private String name;

    private String description;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private ArtifactType artifactType;

    /**
     * Id of the concrete thing this artifact is: a base-entity entity name when
     * {@link #artifactType} is {@link ArtifactType#ENTITY}, a base-document document name when
     * {@link ArtifactType#DOCUMENT}, or a widget name when {@link ArtifactType#WIDGET}.
     */
    private String artifactTypeId;

    /**
     * ID of the state machine definition in base-state that governs this artifact's lifecycle.
     * Passed when publishing {@code ArtifactInstanceCreatedEvent} so base-state can initialize
     * the right machine.
     */
    private String stateMachineId;

    @Version
    private Long version;
}
